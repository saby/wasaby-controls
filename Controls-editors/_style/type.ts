
import {
    ObjectType,
} from 'Types/meta';
import StyleEditor from './editor';

/**
 * Метатип стилей
 */
export const StyleType = ObjectType.editor(() => Promise.resolve(StyleEditor));
StyleType.getEditor().load();
