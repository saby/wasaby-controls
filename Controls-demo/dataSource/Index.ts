import { Control, TemplateFunction } from 'UI/Base';
import * as Template from 'wml!Controls-demo/dataSource/Index';

export default class extends Control {
    protected _template: TemplateFunction = Template;
}
