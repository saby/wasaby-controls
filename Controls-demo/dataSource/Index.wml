<div class="controlsDemo__wrapper">
    <div class="controlsDemo__cell">
        <Router.router:Reference
           attr:class="controls-text-label"
           state="/Controls-demo/app/:app"
           app="Controls-demo/dataSource/Tree">
            <a href="{{content.href}}">Обработка ошибок в дереве</a>
        </Router.router:Reference>
    </div>
</div>
