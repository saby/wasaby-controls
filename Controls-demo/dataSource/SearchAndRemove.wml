<div>
   <div class="controlsDemo__flex controlsDemo_fixedWidth500 controlsDemo__flex-jc-sb">
      <Controls.checkbox:Checkbox
         caption="Разорвать соединение"
         bind:value="_lostConnection"
      />
      <Controls.buttons:Button
         caption="Сбросить source"
         on:click="_resetSource()"
      />
   </div>
   <Controls.browser:Browser
      keyProperty="id"
      source="{{ _source }}"
      navigation="{{ _navigation }}"
      searchParam="title"
      bind:filter="_filter"
      attr:class="controlsDemo_fixedWidth500"
   >
      <div>
            <Controls.search:InputContainer>
               <Controls.search:Input />
            </Controls.search:InputContainer>
            <Controls.list:Container itemActions="{{ _itemActions }}">
               <Controls.scroll:Container attr:style="height: 300px;">
                  <Controls.list:View />
               </Controls.scroll:Container>
            </Controls.list:Container>
            <Controls.listDeprecate:Remover name="listRemover" on:afterItemsRemove="_afterItemsRemove()" />
         </div>
   </Controls.browser:Browser>
</div>
