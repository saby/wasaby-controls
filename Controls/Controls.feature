<?xml version="1.0" encoding="UTF-8" ?>
<features version="1.00">

  <feature id="menu_mode_selector" last_changed="Золотова Э.Е." last_changed_uuid="55dd1897-4ba4-4c0e-b17e-e05fc0a86623" project="https://online.sbis.ru/opendoc.html?guid=22249abd-c279-4400-a71a-a1b986bf2a0c&amp;client=3" responsible="Золотова Э.Е." responsible_uuid="55dd1897-4ba4-4c0e-b17e-e05fc0a86623">
    <description>Включает отображение меню в виде справочника</description>
  </feature>

  <feature id="new-button-shadow" last_changed="Мочалов М.А." last_changed_uuid="f6fdf353-3ab3-4dbb-8671-01c2b418bf0a" project="https://project.sbis.ru/uuid/58c5e7d5-bbd5-498e-8468-a7495197c9ba/page/project-main" responsible="Мочалов М.А." responsible_uuid="f6fdf353-3ab3-4dbb-8671-01c2b418bf0a">
    <description>Включает новое отображение тени кнопки</description>
  </feature>

</features>
