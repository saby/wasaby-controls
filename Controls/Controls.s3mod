<?xml version="1.0" encoding="UTF-8" ?>
<ui_module id="f6fdf353-3ab3-4dbb-8671-01c2b418bf0a" kaizen_zone="916d4071-6bb4-4800-b3ff-2ee6725c9fdc"  name="Controls" package="Wasaby" responsible="Крайнов Д.О." responsible_uuid="0a464b92-3129-49f2-8a5e-7ce6ecfa37b0" required="1" version="1.00">

  <description>Готовый набор визуальных компонентов, отвечающих корпоративным стандартам. Позволяют быстро спроектировать визуальную составляющую веб-приложения.</description>

    <depends>
      <module id="686f183d-0bcc-4732-b4e4-6016c2a67a0f" name="Types"/>
      <module id="323fd54f-1fab-42a6-8aea-6726d6392d53" name="Application"/>
      <module id="63e1a7e5-4d75-42fa-91a3-977dc90ff1e1" name="SbisEnv"/>
      <module id="0df0fce3-53ba-47f8-987d-edf1fd078295" name="WS.Core"/>
      <module id="636177e5-4d75-44fa-9173-977dc90ff111" name="Env"/>
      <module id="4220f331-8217-437b-810c-8c4479323a27" name="EnvConfig"/>
      <module id="2274d173-fe30-4fcf-81ca-767915919d69" name="EnvTouch"/>
      <module id="656177e2-4d75-42fa-9173-977dc90ff121" name="Browser"/>
      <module id="10efeac7-0f02-4c73-9474-0d1d0db077b6" name="View"/>
      <module id="0f5d26db-dd61-4957-9375-8861ad66f219" name="Vdom"/>
      <module id="f27abe12-19df-4c9f-b886-634bd20a0011" name="UI"/>
      <module id="832cda0a-e19c-4ffc-a1d1-47edbb6b7501" name="I18n"/>
      <module id="37e402f1-2fb0-4b73-81c5-113cf94dbf1c" name="WasabyLoader"/>
      <module id="e4bbd0f2-f7ce-404e-88b2-5130d105b770" name="ParametersWebAPI"/>
      <module id="6ae4f1d6-9ecf-4805-a98c-65e47906eedd" name="Tailwind"/>
      <module id="50f80520-0983-4049-b17a-d893bce91d69" name="ErrorHandling"/>
      <module id="8298f3a2-1964-11ed-861d-0242ac120002" name="Controls-DataEnv"/>
    </depends>

  <load_after/>

  <parameters/>

  <resources/>

  <preload/>

</ui_module>
