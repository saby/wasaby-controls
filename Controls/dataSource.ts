/**
 * @kaizen_zone 997e2040-c20b-4857-8580-c283c4b85f85
 */
/**
 * Библиотека компонентов для упрощения загрузки данных: формирования запросов, обработки ошибок.
 * Если нужно обработать ошибки, то вместо этой библиотеки нужно использовать библиотеку {@link Controls/error}.
 * @library
 * @includes error Controls/dataSource:error
 * @includes requestDataUtil Controls/_dataSource/requestDataUtil
 * @public
 */
import * as error from 'Controls/_dataSource/error';
import requestDataUtil from 'Controls/_dataSource/requestDataUtil';
import { nodeHistoryUtil } from 'Controls/_dataSource/nodeHistoryUtil';
import groupUtil from 'Controls/_dataSource/GroupUtil';
import { IDataLoaderOptions } from './_dataSource/DataLoader';
import { ILoadDataConfig } from './_dataSource/DataLoader/interface/ILoadDataConfig';
import { ILoadDataResult } from './_dataSource/DataLoader/interface/ILoadDataResult';
import { ILoadDataCustomConfig } from './_dataSource/DataLoader/interface/ILoadDataCustomConfig';

export {
    error,
    requestDataUtil,
    groupUtil,
    nodeHistoryUtil,
    IDataLoaderOptions,
    ILoadDataConfig as ISourceConfig,
    ILoadDataResult as IRequestDataResult,
};
export { CrudWrapper } from 'Controls/_dataSource/CrudWrapper';
export {
    default as NewSourceController,
    IControllerState as ISourceControllerState,
    IControllerOptions as ISourceControllerOptions,
} from './_dataSource/Controller';
export {
    default as calculatePath,
    Path,
} from 'Controls/_dataSource/calculatePath';
export { isEqualItems } from './_dataSource/Controller';
export { default as NavigationController } from './_dataSource/NavigationController';
export {
    default as DataLoader,
    TLoadersConfigsMap as TLoadConfig,
    TLoadResultMap,
} from './_dataSource/DataLoader';
export { ILoadDataConfig, ILoadDataResult, ILoadDataCustomConfig };
export {
    default as PageController,
    IPageConfig,
} from './_dataSource/PageController';
export {
    getState as getControllerState,
    saveState as saveControllerState,
    IListSavedState,
} from 'Controls/_dataSource/Controller/State';
export const FILTER_EDITORS_WHICH_REQUIRED_DATA_LOAD = [
    'Controls/filterPanel:ListEditor',
    'Controls/filterPanel:LookupEditor',
];
