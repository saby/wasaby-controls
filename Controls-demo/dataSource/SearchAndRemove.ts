import { Control, TemplateFunction } from 'UI/Base';
import { Memory, Query, DataSet } from 'Types/source';
import { adapter, Model } from 'Types/entity';
import { fetch } from 'Browser/Transport';
import * as template from 'wml!Controls-demo/dataSource/SearchAndRemove';
import { Remover } from 'Controls/list';
import { IItemAction } from 'Controls/itemActions';
import { Confirmation } from 'Controls/popup';
import {
    INavigationOptionValue,
    INavigationPageSourceConfig,
} from 'Controls/interface';

interface IFilter {
    title?: string;
}

enum ErrorType {
    serviceError = 502,
    connection = 0,
}

const getStandardError = (httpError: ErrorType = ErrorType.serviceError) => {
    return Promise.reject(
        new fetch.Errors.HTTP({ url: '', httpError, message: '' })
    );
};

class TestSource extends Memory {
    private _lostConnection: boolean = false;

    query(filter: Query<IFilter>): Promise<DataSet> {
        const where = filter.getWhere() as unknown as IFilter;

        if (this._lostConnection) {
            return getStandardError(ErrorType.connection);
        }

        if (where.title === 'error') {
            return getStandardError(ErrorType.serviceError);
        }

        return super.query(filter);
    }

    destroy(): Promise<null> {
        return getStandardError();
    }

    loseConnection(lose: boolean = false): void {
        this._lostConnection = lose;
    }
}

function getSourceData(): object[] {
    const count = 100;
    const data = [];

    for (let i = 0; i < count; i++) {
        data.push({ id: String(i), title: 'Element_' + i });
    }

    return data;
}

function getSource(): TestSource {
    return new TestSource({
        keyProperty: 'id',
        data: getSourceData(),
        filter: (item: adapter.IRecord, { title: desired }: IFilter) => {
            return desired
                ? item.get('title').toLowerCase().indexOf(desired) !== -1
                : true;
        },
    });
}

export default class extends Control {
    protected _template: TemplateFunction = template;
    protected _filter: IFilter = {};
    protected _lostConnection: boolean = false;
    protected _source: TestSource = getSource();

    protected _navigation: INavigationOptionValue<INavigationPageSourceConfig> =
        {
            source: 'page',
            sourceConfig: { hasMore: false, page: 5, pageSize: 10 },
            view: 'infinity', // infinity, pages, demand
            viewConfig: {
                pagingMode: 'basic',
                showEndButton: true,
            },
        };

    protected _itemActions: IItemAction[] = [
        {
            id: '1',
            icon: 'icon-Erase',
            iconStyle: 'danger',
            showType: 2,
            title: 'Remove',
            handler: (item: Model) => {
                this._children.listRemover.removeItems([item.getKey()]);
            },
        },
    ];

    protected _children: {
        listRemover: typeof Remover;
    };

    protected _afterUpdate(): void {
        this._source.loseConnection(this._lostConnection);
    }

    protected _resetSource(): void {
        this._source = getSource();
        this._lostConnection = false;
    }

    protected _afterItemsRemove(event: unknown, idArray: string[]): boolean {
        if (idArray[0] !== '0') {
            return true;
        }

        Confirmation.openPopup({
            message: 'Нельзя удалить первый элемент!',
            style: 'danger',
            type: 'ok',
        });

        return false;
    }
}
