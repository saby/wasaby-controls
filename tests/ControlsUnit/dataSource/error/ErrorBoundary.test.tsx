/**
 * @jest-environment jsdom
 */
import { unmountComponentAtNode } from 'react-dom';
import { render } from '@testing-library/react';
import { error } from 'Controls/dataSource';
import * as ErrorHandling from 'ErrorHandling/ErrorHandling';
import type { ErrorViewConfig, IErrorHandlerConfig } from 'ErrorHandling/interface';

const { ErrorBoundary } = error;

interface IHelloWorldProps {
    hasError?: true;
}

function HelloWorld({ hasError }: IHelloWorldProps): JSX.Element {
    if (hasError) {
        throw new Error('custom error');
    }

    return <div>Hello world!</div>;
}

interface IErrorTemplateProps {
    message: string;
}

function ErrorTemplate({ message }: IErrorTemplateProps): JSX.Element {
    return <div>This is error template: {message}</div>;
}

function errorHandler({ error }: IErrorHandlerConfig): ErrorViewConfig {
    return {
        template: ErrorTemplate,
        options: {
            message: error.message,
        },
    };
}

describe('Controls/dataSource:error.ErrorBoundary', () => {
    let container: HTMLDivElement;
    let originalErrorHandlers;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
        originalErrorHandlers = ErrorHandling.getHandlers();
        ErrorHandling.setHandlers([errorHandler]);
    });

    afterEach(() => {
        unmountComponentAtNode(container);
        container.remove();
        ErrorHandling.setHandlers(originalErrorHandlers);
        container = null;
        originalErrorHandlers = null;
    });

    it('Рендер компонента без ошибки', () => {
        render(
            <ErrorBoundary>
                <HelloWorld />
            </ErrorBoundary>,
            { container }
        );

        expect(container).toMatchSnapshot();
    });

    it('Рендер компонента с ошибкой', () => {
        render(
            <ErrorBoundary>
                <HelloWorld hasError={true} />
            </ErrorBoundary>,
            { container }
        );

        expect(container).toMatchSnapshot();
    });
});
