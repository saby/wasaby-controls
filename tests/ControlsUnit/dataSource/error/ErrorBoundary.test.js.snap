// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls/dataSource:error.ErrorBoundary Рендер компонента без ошибки 1`] = `
<div>
  <div>
    Hello world!
  </div>
</div>
`;

exports[`Controls/dataSource:error.ErrorBoundary Рендер компонента с ошибкой 1`] = `
<div>
  <div
    class="tw-contents"
  >
    <div>
      This is error template: 
      custom error
    </div>
  </div>
</div>
`;
